### ISSUE TITLE
Add workspace method to keep sources

### ISSUE THREAD

**Issue by cle1109** [scot-dev/scot#141]
In addition to `ws.remove_sources`, we should have a `ws.keep_sources`.

**Comment by cle1109** [comment]
In addition, would it make sense not to manipulate the matrices directly, but to have a mask that marks the sources that should be used in all computations?

**Comment by mbillingr** [comment]
Yes, this makes sense.

I think we would need getter methods for the matrices to relieve the user from dealing with the masks.

**Comment by cle1109** [comment]
Do we? All functions working with the matrices could automatically apply the mask.

**Comment by mbillingr** [comment]
If users want to get the matrices they should not have to apply the mask themselves.

**Comment by cle1109** [comment]
But what if they want the full non-masked matrices?

**Comment by mbillingr** [comment]
Nobody forces them to remove components, first ;)
And they can access the class attributes, directly too.

**Comment by cle1109** [comment]
OK. Then we need a method to reset the mask, e.g. `ws.reset_components`. I assume we will only modify that for the `Workspace` class?

Which variables do you want to decorate with `@property`? How should we modify the names? This could get cumbersome if we call the property `self.mixing_` and the corresponding private variable `self._mixing_` - don't you think?

**Comment by mbillingr** [comment]
I had been thinking in simpler terms of implementing `get_mixing()` and `get_unmixing()` methods.

**Comment by cle1109** [comment]
That's so Java and not Pythonic :-). So you think we need to consider `mixing` and `unmixing`. What about `activations?` Anything else?

**Comment by mbillingr** [comment]
That's not Java, that's C++ :)
I can't think of anything else at the moment.

**Comment by cle1109** [comment]
Well, why don't we use the decorators then?

**Comment by mbillingr** [comment]
I never said we shouldn't use `@property`. Properties are rather nice and we can use them, here. (And they are simpler to use than I thought.)

Nevertheless, I am surprised that properties are supported in Python. I always thought that one of the principles of being Pythonic is *explicit is better than implicit*. With a property the user thinks they just access a value stored in the class, but actually it involves a function call where the value is computed on the fly. A getter function would be much more explicit, and therefore much more Pythonic, no?

**Comment by cle1109** [comment]
> I never said we shouldn't use @property. Properties are rather nice and we can use them, here. (And they are simpler to use than I thought.)

OK, then let's go for it.

I like properties because they are nothing special in Python - they're just decorators (which is just syntactic sugar for wrapping functions). I don't know if using properties is implicit - after all, I don't think that users think they're accessing a value. In fact, _there should be one - and preferably only one -obvious way to do it._ Being able to access a value via `object.value` and `object.get_value()` is certainly more than one way.

Anyway, this is getting philosophical - which is great, but I think we should discuss it outside of this issue :-). Here, we should be pragmatic and decide what to do. It looks like you could be onboard with properties, right?

**Comment by mbillingr** [comment]
> It looks like you could be onboard with properties, right?

Yes. In order to use them I don't have to understand why it is pythonic to use them :)

**Comment by cle1109** [comment]
> Yes. In order to use them I don't have to understand why it is pythonic to use them :)

Absolutely :-).

I found this blog post, which although it contains a few gotchas regarding the use of class variables, it does explain why properties are Pythonic: http://blaag.haard.se/What-s-the-point-of-properties-in-Python/
